Refactor Apparmor Profile generation

daemon1024 opened the pull request:
We now have a way to resolve conflicts between local and global profile allowing us to set a priority order in the code for what to expect when conflicting rules exist between a global policy and fromSource.

This conflict deemed a complex resolution logic so I ended up refactoring the profile generation logic to more freely analyse and manipulate the profiles.

Will handle in BPF LSM in separate PR

Ref #787

daemon1024 committed:
refactor apparmor profile generation using templates

templating here help us drop in values inside a template without needing to manually generate individual apparmor rule or line

this would also provide us with the ability to freely analyse the rule entities without needing apply various string manipulations

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
resolve local-global profile conflict

This commit provides a way to resolve conflicts between local and global profile allowing us to set a priority order for what to expect when conflicting rules exist between a global policy and fromSource policy

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul commented on KubeArmor/enforcer/appArmorTemplate.go:
```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 Authors of KubeArmor
```
```suggestion
// Copyright 2022 Authors of KubeArmor
```

nyrahul reviewed (commented):
This is a Photogenic PR! LGTM

daemon1024 committed:
keep deny and allow lists seperate

Deny and Allow are now not either or but can coexist, we leave the conflict resolution to AppArmor

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
test: add test case to demonstrate lenient whitelist policy

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul reviewed (approved):
LGTM
